src/libexpr/primops/fetchTree.cc

```diff
@@ -182,45 +182,208 @@ static RegisterPrimOp primop_fetchTree({
-      The input should be "locked", that is, it should contain a commit hash or content hash unless impure evaluation (`--impure`) is enabled.
+      Fetch a file system tree or a plain file using one of the supported backends and return a [store path](@docroot@/glossary.md#gloss-store-path) with metadata.
 
-      > **Note**
-      >
-      > The URL-like syntax requires the [`flakes` experimental feature](@docroot@/contributing/experimental-features.md#xp-feature-flakes) to be enabled.
+      *input* must be an attribute set with the following attributes:
 
-      Here are some examples of how to use `fetchTree`:
+      - `type` (String, required)
 
-      - Fetch a GitHub repository using the attribute set representation:
+        Source type.
+        This determines other allowed input attributes.
+        The next section lists source types and associated parameters.
 
-          ```nix
-          builtins.fetchTree {
-            type = "github";
-            owner = "NixOS";
-            repo = "nixpkgs";
-            rev = "ae2e6b3958682513d28f7d633734571fb18285dd";
-          }
-          ```
+      - `narHash` (String, optional)
 
-        This evaluates to the following attribute set:
+        The `narHash` parameter can be used to substitute the source of the tree.
+        It also allows for verification of tree contents that may not be provided by the underlying transfer mechanism.
+        If `narHash` is set, the source is first looked up is the Nix store and [substituters](@docroot@/command-ref/conf-file.md#conf-substituters), and only fetched if not available.
 
-          ```
-          {
-            lastModified = 1686503798;
-            lastModifiedDate = "20230611171638";
-            narHash = "sha256-rA9RqKP9OlBrgGCPvfd5HVAXDOy8k2SmPtB/ijShNXc=";
-            outPath = "/nix/store/l5m6qlvfs9sdw14ja3qbzpglcjlb6j1x-source";
-            rev = "ae2e6b3958682513d28f7d633734571fb18285dd";
-            shortRev = "ae2e6b3";
-          }
-          ```
+      A subset of the output attributes of `fetchTree` can be re-used for subsequent calls to `fetchTree` to produce the same result again.
+      That is, `fetchTree` is idempotent.
 
-      - Fetch the same GitHub repository using the URL-like syntax:
+      Sources not secured by `narHash` are cached in `$XDG_CACHE_HOME/nix`.
+      The cache timeout is determined by [`tarball-ttl`](@docroot@/command-ref/conf-file.md#conf-tarball-ttl).
 
-          ```
-          builtins.fetchTree "github:NixOS/nixpkgs/ae2e6b3958682513d28f7d633734571fb18285dd"
-          ```
+      ## Source types
+
+      The following source types and associated input attributes are supported.
+
+      <!-- TODO: It would be soooo much more predictable to work with (and
+      document) if `fetchTree` was a curried call with the first paramter for
+      `type` or an attribute like `builtins.fetchTree.git`! -->
```

Comment:
Very interesting, nested builtin to disambiguation the type. On the other hand, this makes the idemlotency a little more awkward.